### PR TITLE
test(DRIVERS-2657): break out lambda scripts 

### PR DESCRIPTION
**Pull request by durran**
Breaks out the scripts that run the lambda tests into multiple scripts, with the atlas create/delete cluster scripts broken out so that they can be run in setup/teardown on evergreen task groups.

Node.js impl: https://github.com/mongodb/node-mongodb-native/pull/3740
Spec changes: https://github.com/mongodb/specifications/pull/1438
Original script: For context, the original script is https://github.com/mongodb-labs/drivers-evergreen-tools/blob/master/.evergreen/run-deployed-lambda-aws-tests.sh

Example evergreen config:

```yml
tasks:
  - name: "test-aws-lambda-deployed"
    commands:
      - func: "install dependencies"
      - command: ec2.assume_role
        params:
          role_arn: ${LAMBDA_AWS_ROLE_ARN}
          duration_seconds: 3600
      - command: subprocess.exec
        params:
          working_dir: src
          binary: bash
          add_expansions_to_env: true
          args:
            - ${DRIVERS_TOOLS}/.evergreen/aws_lambda/run-deployed-lambda-aws-tests.sh
          env:
            TEST_LAMBDA_DIRECTORY: ${PROJECT_DIRECTORY}/test/lambda
            AWS_REGION: us-east-1
task_groups:
  - name: test_aws_lambda_task_group
    setup_group:
      - func: fetch source
      - command: subprocess.exec
        params:
          working_dir: src
          binary: bash
          add_expansions_to_env: true
          args:
            - ${DRIVERS_TOOLS}/.evergreen/atlas/setup-atlas-cluster.sh
      - command: expansions.update
        params:
          file: src/atlas-expansion.yml
    teardown_group:
      - command: subprocess.exec
        params:
          working_dir: src
          binary: bash
          add_expansions_to_env: true
          args:
            - ${DRIVERS_TOOLS}/.evergreen/atlas/teardown-atlas-cluster.sh
    setup_group_can_fail_task: true
    setup_group_timeout_secs: 1800
    tasks:
      - test-aws-lambda-deployed
```

### CODE REVIEW

File: .evergreen/aws_lambda/run-deployed-lambda-aws-tests.sh [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    echo "No cluster could be created in the 15 minute timeframe or error occured."
+    exit 1
+  else
+    echo "Setting MONGODB_URI in the environment to the new cluster."
+    # else set the mongodb uri
+    URI=$(echo $SRV_ADDRESS | grep -Eo "[^(\/\/)]*$" | cat)
+    MONGODB_URI="mongodb+srv://${DRIVERS_ATLAS_LAMBDA_USER}:${DRIVERS_ATLAS_LAMBDA_PASSWORD}@${URI}"
+    export MONGODB_URI=$MONGODB_URI
+  fi
+}
```
Comment (baileympearson) [inline .evergreen/aws_lambda/run-deployed-lambda-aws-tests.sh:59]: Does this logic belong in the setup script instead?  I'd expect the setup script to fail if its unable to provision an atlas cluster in the time limit (15 minutes).  

I'm thinking of reusing the setup/teatdown (with potentially some slight modifications) for the search indexes testing, but having the `check_cluster` logic inside the script to run the lambda tests makes it hard to share the logic.

Comment (durran) [reply .evergreen/aws_lambda/run-deployed-lambda-aws-tests.sh:59]: Oh I'm still in draft here. :) I was planning on moving it to the setup script - just needed to get it working then figure out the best way to get the MONGODB_URI passed through.

Comment (baileympearson) [reply .evergreen/aws_lambda/run-deployed-lambda-aws-tests.sh:59]: Sorry to step on your toes during draft phase, should have looked more closely at the PR before commenting 😬

Comment (durran) [reply .evergreen/aws_lambda/run-deployed-lambda-aws-tests.sh]: Updated to move the check into the setup and write the URI to a YML file for expansion usage in evg.

**Review by baileympearson — APPROVED** [review]

File: .evergreen/aws_lambda/setup-atlas-cluster.sh [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+# Explanation of required environment variables:
+#
+# DRIVERS_ATLAS_PUBLIC_API_KEY: The public Atlas key for the drivers org.
+# DRIVERS_ATLAS_PRIVATE_API_KEY: The private Atlas key for the drivers org.
+# DRIVERS_ATLAS_GROUP_ID: The id of the individual projects under the drivers org, per language.
+# DRIVERS_ATLAS_LAMBDA_USER: The user for the lambda cluster.
+# DRIVERS_ATLAS_LAMBDA_PASSWORD: The password for the user.
+# LAMBDA_STACK_NAME: The name of the stack on lambda "dbx-<language>-lambda"
+# MONGODB_VERSION: The major version of the cluster to deploy.
```
Comment (baileympearson) [inline .evergreen/aws_lambda/setup-atlas-cluster.sh:12]:
```suggestion
# MONGODB_VERSION: The major version of the cluster to deploy.  Defaults to 6.0.
```

**Comment by blink1073** [comment]
Does the lambda stack deletion need to be part of teardown as well?

File: .evergreen/aws_lambda/teardown-atlas-cluster.sh [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+# LAMBDA_STACK_NAME: The name of the stack on lambda "dbx-<language>-lambda"
+
+# Explanation of generated variables:
+#
+# FUNCTION_NAME: Uses the stack name plus the current commit sha to create a unique cluster and function.
+# ATLAS_BASE_URL: Where the Atlas API root resides.
+
+# The base Atlas API url. We use the API directly as the CLI does not yet
+# support testing cluster outages.
+ATLAS_BASE_URL="https://cloud.mongodb.com/api/atlas/v1.0"
```
Comment (prestonvasquez) [inline .evergreen/aws_lambda/teardown-atlas-cluster.sh:18]: Should we decouple the version in the ATLAS_BASE_URL?

```suggestion
ATLAS_API_VERSION="v.10"
ATLAS_BASE_URL="https://cloud.mongodb.com/api/atlas/$ATLAS_API_VERSION"
```

File: .evergreen/aws_lambda/setup-atlas-cluster.sh [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+# Explanation of generated variables:
+#
+# MONGODB_URI: The URI for the created Atlas cluster during this script.
+# FUNCTION_NAME: Uses the stack name plus the current commit sha to create a unique cluster and function.
+# CREATE_CLUSTER_JSON: The JSON used to create a cluster via the Atlas API.
+# ATLAS_BASE_URL: Where the Atlas API root resides.
+
+# The base Atlas API url. We use the API directly as the CLI does not yet
+# support testing cluster outages.
+ATLAS_BASE_URL="https://cloud.mongodb.com/api/atlas/v1.0"
```
Comment (prestonvasquez) [inline .evergreen/aws_lambda/setup-atlas-cluster.sh:23]: Should we decouple the version in the ATLAS_BASE_URL?

```suggestion
ATLAS_API_VERSION="v.10"
ATLAS_BASE_URL="https://cloud.mongodb.com/api/atlas/$ATLAS_API_VERSION"
```

File: .evergreen/aws_lambda/teardown-atlas-cluster.sh (hunk 3 above)
Comment (durran) [reply .evergreen/aws_lambda/teardown-atlas-cluster.sh]: Done.

File: .evergreen/aws_lambda/setup-atlas-cluster.sh (hunk 4 above)
Comment (durran) [reply .evergreen/aws_lambda/setup-atlas-cluster.sh]: Done.

File: .evergreen/aws_lambda/run-deployed-lambda-aws-tests.sh [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+set -o errexit  # Exit the script with error if any of the commands fail
+
+# Explanation of required environment variables:
+#
+# TEST_LAMBDA_DIRECTORY: The root of the project's Lambda sam project.
+# DRIVERS_ATLAS_PUBLIC_API_KEY: The public Atlas key for the drivers org.
+# DRIVERS_ATLAS_PRIVATE_API_KEY: The private Atlas key for the drivers org.
+# DRIVERS_ATLAS_GROUP_ID: The id of the individual projects under the drivers org, per language.
+# LAMBDA_STACK_NAME: The name of the stack on lambda "dbx-<language>-lambda"
+# AWS_REGION: The region for the function - generally us-east-1
```
[the remark this answers is not in the document]
Comment (durran) [reply .evergreen/aws_lambda/run-deployed-lambda-aws-tests.sh:11]: Added `MONGODB_URI` as a required env variable in the comments.

**Comment by durran** [comment]
> Does the lambda stack deletion need to be part of teardown as well?

@blink1073 I didn't split out lambda cloud formation deploy/delete into their own setup/teardown as I wanted it to be apparent in the main task that it was a lambda flow with the assume role and running of the script, with the atlas related tasks being split out. I can split those out as well if you feel strongly about it.

File: .evergreen/aws_lambda/setup-atlas-cluster.sh [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+  echo "Creating new Atlas Cluster..."
+  curl \
+    --digest -u "${DRIVERS_ATLAS_PUBLIC_API_KEY}:${DRIVERS_ATLAS_PRIVATE_API_KEY}" \
+    -d "${CREATE_CLUSTER_JSON}" \
+    -H 'Content-Type: application/json' \
+    -X POST \
+    "${ATLAS_BASE_URL}/groups/${DRIVERS_ATLAS_GROUP_ID}/clusters?pretty=true"
+}
+
+# Check is cluster has a srv address, and assume once it does, it can be used.
```
Comment (alcaeus) [inline .evergreen/aws_lambda/setup-atlas-cluster.sh:93]:
```suggestion
# Check if cluster has a srv address, and assume once it does, it can be used.
```

File: .evergreen/aws_lambda/setup-atlas-cluster.sh [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+check_cluster ()
+{
+  count=0
+  SRV_ADDRESS="null"
+  # Don't try longer than 20 minutes.
```
Comment (alcaeus) [inline .evergreen/aws_lambda/setup-atlas-cluster.sh:98]: What's the average time we need to complete this? Waiting for 20 minutes seems like an awfully long time, would it be safe to shorten this to 5 or 10 minutes?

Comment (durran) [reply .evergreen/aws_lambda/setup-atlas-cluster.sh:98]: Since we need a M10 at minimum to test failover, this can take some time. I've seen it over 15 minutes in cases but haven't seen it over 20, so I set the timeout at 20 here.

Comment (alcaeus) [reply .evergreen/aws_lambda/setup-atlas-cluster.sh:98]: Ah, I see. Thanks for explaining!

**Review by prestonvasquez — APPROVED** [review]

**Review by ShaneHarvey — APPROVED** [review]
